executor/src/witgen/generator.rs

```diff
@@ -227,43 +241,35 @@ impl<'a, T: FieldElement> Generator<'a, T> {
                 "  Checking that remaining identities hold when unknown values are set to 0"
             );
             self.process_identities(
+                row_index,
                 &mut identities_without_next_ref,
                 UnknownStrategy::Zero,
                 &mut identity_processor,
             )
             .and_then(|_| {
                 self.process_identities(
+                    row_index,
                     &mut identities_with_next_ref,
                     UnknownStrategy::Zero,
                     &mut identity_processor,
                 )
             })
-            .map_err(|e| self.report_failure_and_panic_underconstrained(e))
+            .map_err(|e| self.report_failure_and_panic_underconstrained(row_index, e))
             .unwrap();
         }
 
         log::trace!(
             "{}",
-            self.current.render(
-                &format!("===== Row {}", self.current_row_index),
-                true,
-                &self.witnesses
-            )
+            self.row(row_index)
+                .render(&format!("===== Row {}", row_index), true, &self.witnesses)
         );
-
-        self.shift_rows();
-
-        if next_row == 0 {
-            self.first = self.previous.clone()
-        }
-
-        self.previous.clone().into()
     }
 
     /// Loops over all identities and queries, until no further progress is made.
     /// @returns the "incomplete" identities, i.e. identities that contain unknown values.
     fn loop_until_no_progress<Q>(
         &mut self,
+        row_index: DegreeType,
```

Comment:
I'm wondering if it would be simpler to extract some part that can have `row_index` as a member of `self`, but just a suggestion.

Reply:
Hmm, before this PR, the current row index was a member of `Generator`. I didn't think that was clean, because it's not obvious where it is updated.

I think what you're suggesting is to have a new type that's only instantiated for a particular row, so it would stay constant for the lifetime of that object, right? I think that could be a future refactoring when we try to get a generic `Processor`.